### PR TITLE
igl | android | Support sRGB framebuffer on OpenGLES.

### PR DESCRIPTION
**Pull request by vinsentli**
image

### CODE REVIEW

File: shell/android/jni/TinyRenderer.cpp [hunk 1]
```diff
@@ -100,6 +100,16 @@ void TinyRenderer::init(AAssetManager* mgr,
     vulkan::VulkanContextConfig config;
     config.terminateOnValidationError = true;
     config.requestedSwapChainTextureFormat = swapchainColorTextureFormat;
+    switch (swapchainColorTextureFormat) {
```
Comment (francoiscoulombe) [inline shell/android/jni/TinyRenderer.cpp:103]: Hi, while I believe you that you get different results, why do you feel this is the right fix?

can you provide screenshots of the differences just to make sure that we’re talking about the same thing?

the sRGB frame buffer might not be very well tested on Android phones

**Comment by vinsentli** [comment]
@francoiscoulombe You can check out these screenshots
| Header | Header | Header |
|--------|--------|--------|
| OpenGL BGRA_SRGB & RGBA_UNorm8  | Vulkan before : BGRA_SRGB | Vulkan  now :RGBA_UNorm8 |
| d380a7a8db98d5eb675b3037d8d6a7e0_compress | a6d1473835817baceee6782fdc9a7b36_compress | 1d3f677a388fb1b63d7f76706b030e94_compress |

**Comment by francoiscoulombe** [comment]
thanks for the screenshots.

for what it's worth, the center one is the correct colors

I don't know why sending a BGRA_SRGB to OpenGL ES isn't resulting in an image that is more like the center one.

I'm more worried about the swapchain color space changes which are separate from this

**Comment by francoiscoulombe** [comment]
to verify set your display into srgb profile and sample the pixel to compare with the value written on the image.
image

**Comment by vinsentli** [comment]
Indeed, the color in the middle image is correct.

**Comment by vinsentli** [comment]
I will continue researching the reasons behind the incorrect colors in OpenGL.

**Comment by vinsentli** [comment]
@francoiscoulombe I have resolved it.
image

**Comment by francoiscoulombe** [comment]
What was it?

**Comment by vinsentli** [comment]
We need create a sRGB surface in GLSurfaceView.

The prerequisite is that the device needs to support the EGL_KHR_gl_colorspace extension.

```java

setEGLWindowSurfaceFactory(new SurfaceFactory(getHolder()));

private static class SurfaceFactory implements GLSurfaceView.EGLWindowSurfaceFactory{
    private SurfaceHolder surfaceHolder_;

    public SurfaceFactory(SurfaceHolder surfaceHolder){
      surfaceHolder_ = surfaceHolder;
    }

    final int EGL_GL_COLORSPACE_KHR = 0x309D;
    final int EGL_GL_COLORSPACE_SRGB_KHR = 0x3089;
    final int EGL_GL_COLORSPACE_LINEAR_KHR        =    0x308A;

    final int[] configAttribs = {
            EGL_GL_COLORSPACE_KHR ,EGL_GL_COLORSPACE_SRGB_KHR,
            EGL10.EGL_NONE
    };

    @Override
    public EGLSurface createWindowSurface(EGL10 egl10, EGLDisplay eglDisplay, EGLConfig eglConfig, Object o) {
      return egl10.eglCreateWindowSurface(eglDisplay, eglConfig, surfaceHolder_, configAttribs);
    }

    @Override
    public void destroySurface(EGL10 egl10, EGLDisplay eglDisplay, EGLSurface eglSurface) {
      egl10.eglDestroySurface(eglDisplay, eglSurface);
    }
  }
```

**Comment by francoiscoulombe** [comment]
Sweet!
on Vulkan, you only need the colorspace extension to do a color space that is not nonlinear srgb. (Like pass through)

nonlinear srgb “should” work pretty much everywhere nowadays.

**Comment by vinsentli** [comment]
That's correct.

**Comment by vinsentli** [comment]
After this pull request https://github.com/facebook/igl/pull/199 is merged, I will create a new pull request based on the latest OpenGL SampleView to fix the colorspace issue in OpenGL.

**Comment by francoiscoulombe** [comment]
I’ll take a look first thing on monday

thanks for doing this

**Comment by vinsentli** [comment]
@francoiscoulombe
I have already updated the code.

**Comment by francoiscoulombe** [comment]
sorry for the delay, I had some fire to extinguish

promise I'll try this out today

**Comment by francoiscoulombe** [comment]
using this code, opengl is still dark. 

am i missing something?

**Comment by vinsentli** [comment]
The prerequisite is that the device needs to support the EGL_KHR_gl_colorspace extension.

**Comment by francoiscoulombe** [comment]
you're right... surprisingly, my s22 doesn't seem to have that extension.

gonna try on a different phone

**Comment by vinsentli** [comment]
Please check the format of the swapchain is SRGB.

**Comment by francoiscoulombe** [comment]
i even hardcoded srgb color format in a few places where it was set to unorm8 but unfortunately i'm still getting dark colors

**Comment by vinsentli** [comment]
add a breakpoint in java createWindowSurface,check the final congfig attribute is EGL_GL_COLORSPACE_SRGB_KHR

**Comment by francoiscoulombe** [comment]
i'm not setup with a java debugger but i'm adding logs and they're not coming out

either something's wrong with my setup or something is missing.

I'm also not building this using cmake. maybe i should try to replicate your setup?
how are you running this?

**Comment by corporateshark** [comment]
@francoiscoulombe The files in`build/android/app/src/main/java/com/facebook/igl/sample` are CMake-only. It can be you are using a completely different code path.

**Comment by francoiscoulombe** [comment]
i just saw this. your changes were only done to the files inside the build folder and that's the not the files that i'm using lol

**Comment by francoiscoulombe** [comment]
alright, got it. it works now that i use the right java code. sorry for all the confusion, i didn't know we had this duplication

**Pull request closed without merging**